### ISSUE TITLE
Roadmap to PHP 7.2?

### ISSUE THREAD

**Issue by chilinux** [nextcloud/server-packages#33]
As of NextCloud 15, the Administration Guide recommends using PHP 7.2.

Is there any plans to release an alternative SPEC file which already has the dependencies set to the SCL PHP 7.2 packages?  Otherwise, is there any plans to make the switch to PHP 7.2 package with the release of NextCloud 16 in April?

As a side note/rant: I really wish the SCL would provide the option to update RPM to 4.13 or 4.14 so that boolean dependencies could work on RHEL/CentOS 7.

**Comment by gsanchietti** [comment]
I can't see the recommendation for PHP 7.2, from the [system requirements](https://docs.nextcloud.com/server/stable/admin_manual/installation/system_requirements.html) page supported versions are from 7.0 to 7.3.

There is only a [guide about using 7.2](https://docs.nextcloud.com/server/stable/admin_manual/installation/php_72_installation.html) but not a recommendation.

But SCL PHP 7.1 will be [EOL](https://access.redhat.com/support/policy/updates/rhscl) on Oct 2019, so we need for sure to plan the upgrade.

Still I agree that NextCloud 16 release will be a very good occasion for switching to PHP 7.2!
I don't see any drawback going this way.

Since [NC 16 alpha is out](https://github.com/nextcloud/server/releases/tag/v16.0.0alpha1), do you want to propose a PR for it which uses SCL PHP 7.2?

**Comment by chilinux** [comment]
After putting a lot of through into this, I think you might want to consider breaking the SPEC file into multiple SPEC files.

I purpose moving all the rh-php71 requirements into a new SPEC file called nextcloud-php71 which would have a virtual capability/provides of nextcloud-php (without any version).

There could then also be a nextcloud-php72 with rh-php72 requirements which also has the provides line of nextcloud-php.

The primary SPEC file would then depend on the virtual capability which would draw in the PHP requirements while allowing the user to choose which PHP version best fits their needs/environment at the immediate moment.

Also, as long as the PHP requirements don't change, the version of the nextcloud-php file would *NOT* need to be bumped every time.  So, if 15.0.5 has the same PHP requirements as 15.0.0 then the nextcloud-php71-15.0.0 could still be used.  If 15.0.6 then does require an additional PHP package, the nextcloud-php could be version bumped at that time and the nextcloud SPEC changed to require it virtual capability be version 15.0.6 or higher.

**Comment by gsanchietti** [comment]
It seems quite complex to me with only very limited benefit.
But I will gladly review your PR.

**Comment by chilinux** [comment]
I would like to eventually get to the point of hosting an yum repository of already built RPMs.  If an user has a Nextcloud plugin that still depends on PHP 7.1 and has to modify the SPEC file to revert the version of PHP in the dependencies, then that would kind of defeat the benefit of having already prebuilt RPMs.

I also figured the additional complexity would pay off in terms of streamlining a migration path as PHP continue to evolve without forcing all users to use the latest version (such as when PHP 7.3 become recommended).

I'm not at the PR stage yet but I do now have an example of the type of direction I was looking to go in.  Nothing I have changed in my fork is written in stone for me and I'm interested in whatever feedback you can give on what I'm doing.

The fork is at: https://github.com/chilinux/server-packages

**Comment by gsanchietti** [comment]
> I would like to eventually get to the point of hosting an yum repository of already built RPMs. If an user has a Nextcloud plugin that still depends on PHP 7.1 and has to modify the SPEC file to revert the version of PHP in the dependencies, then that would kind of defeat the benefit of having already prebuilt RPMs.

I think that plugins must have the same requirements as the core.

There is no official yum repository because Nextcloud doesn't really support the installation from RPMs.
Still, we regularly publish the RPMs on NethServer repositories: http://packages.nethserver.org/nethserver/

> I also figured the additional complexity would pay off in terms of streamlining a migration path as PHP continue to evolve without forcing all users to use the latest version (such as when PHP 7.3 become recommended).

We (@alepolidori  and I) are committed to support only the latest release since we need to assure the quality of hundreds of installations.
If you think the complexity would pay off, I need your commitment on supporting the extra packages for all EL7 lifetime.

> The fork is at: https://github.com/chilinux/server-packages

It's a good starting point, but I don't get why you are bringing config files inside the spec like [here](https://github.com/chilinux/server-packages/commit/461de55e7af3d2b1c7639ee6041ff17d41060834#diff-799c2a73d26c08a5c0886613c61d84b7R68).

Also, make sure to create more commits (one for each step of the refactor), IMO  this will ease the maintainability of work.

**Comment by chilinux** [comment]
Official Nextcloud plugins should have the same PHP requirements as the core, just as you said.

There are several unofficial Nextcloud plugins that fall outside the scope of QA testing by the main development team.  In this case it will be the earlier adopters of newer PHP versions that break the plugin that put pressure on the plugin developers.  So far the PHP developers seem to be minimizing changes which would break existing Nextcloud plugins.  I was just trying to provide a fallback option for any cases in the future when that is not the case.

As to the NethServer repos, these look to be a good resource.  Is there any reason they are not promoted in the Readme for this project?  Also, is there any plans to provide a GPG signing key and to sign each of the RPMs in the repo?

I am not completely sure if supporting optional PHP versions would be worth the complexity.  I am backing off on that change for now to see if anyone else has feedback on if that would help them.

I saw moving the configuration file into the SPEC file as a way to make it easier to minimize the steps a novice would need to take.  It would just involve doing a git clone, spectool to fetch the source and rpmbuild to make the RPM.  Leaving the configuration file outside of the spec mean an additional step of copying it to the rpm-build source directory.  It might be possible to achieve the same goal I was going for by making the spec file reference a URL to the github raw content such as:
https://raw.githubusercontent.com/nextcloud/server-packages/master/centos/nextcloud.conf

I will also try to get better at doing more commits to better show my work/steps.  I completely understand where you are coming from with this feedback.  So far I have largely work on github repos were I am the only major contributor.

**Comment by gsanchietti** [comment]
> As to the NethServer repos, these look to be a good resource. Is there any reason they are not promoted in the Readme for this project? Also, is there any plans to provide a GPG signing key and to sign each of the RPMs in the repo?

We didn't want to sponsor our project, also the repository contains a RPM (squid) which overrides the EL7 default one.

By the way the config for the repo is:
```
[nethserver-updates]
name=NethServer-$nsrelease - Updates
mirrorlist=http://mirrorlist.nethserver.org/?release=$releasever&repo=updates&arch=$basearch&nsrelease=$nsrelease
#baseurl=http://mirror.nethserver.org/nethserver/$releasever/updates/$basearch/
gpgcheck=1
repo_gpgcheck=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-NethServer-$releasever
enabled=1
enablegroups=1
```

RPMs are already signed using [this key](https://github.com/NethServer/nethserver-release/blob/master/root/etc/pki/rpm-gpg/RPM-GPG-KEY-NethServer-7).

> I am not completely sure if supporting optional PHP versions would be worth the complexity. I am backing off on that change for now to see if anyone else has feedback on if that would help them.

I agree :)
 
> I saw moving the configuration file into the SPEC file as a way to make it easier to minimize the steps a novice would need to take. It would just involve doing a git clone, spectool to fetch the source and rpmbuild to make the RPM.

Maybe it's simpler but it's not the best practice from upstream. Take a look to SRPM like postfix, httpd or others: the configuration file is always outside the spec. This approach also help on having a cleaner git history.

Thank you for your effort!

**Comment by chilinux** [comment]
For the NextCloud 16 release that is coming out shortly, are you still the most comfortable with SCL being the provider of the PHP packages?  The reason I ask is that SCL's PHP 7.2 is based around PHP 7.2.10 and built back in December.  Remi repository is based around PHP 7.2.17 which were built in April.  There have been 9 security CVEs released for PHP 7.2 in 2019 so far that do not seem to be patched for SCL's release yet.

I wanted to get feedback on how you felt about either switching or bringing back the nextcloud-php virtual package to allow the end user to choose between SCL or Remi.  The virtual packages would then handle the difference in package naming for the dependencies such as rh-php72-fpm (SCL) vs php-fpm (Remi) thus allowing the option for both to be supported.

If we are sticking with a KISS/simple is better and what is the most important is the quality across hundreds of installs, then I have to admit the improved update cycle of Remi is a major draw for me.  If continuing to support SCL is important then the added complexity I was suggesting before might be worth it (maybe?).

It should also be noted in the readme a recommendation to remove rh-php71 via yum (which will take care of dependencies).  Installing rh-php72 will *NOT* replace rh-php71 as they go into different /opt directories and are designed to be installable at the same time.  What is not addressed for dual version installs is which version gets a monopoly on binding to TCP Port 9000 for the FPM.  Someone may be under the impression they have upgraded by installing rh-php72 but actually still be using the rh-php71-fpm package.

Lastly, is there any plans to create a v15 branch once NextCloud 16 is in production release?

Thanks!

**Comment by gsanchietti** [comment]
> For the NextCloud 16 release that is coming out shortly, are you still the most comfortable with SCL being the provider of the PHP packages?

I'd say yes, I would like to keep everything much closer as possible to upstream. Remi is part of [SCLo SIG](https://wiki.centos.org/SpecialInterestGroup/SCLo) and his works usually flows from his personal repositories to upstream ones. 
RHEL will maintain PHP 7.2 [until next year](https://access.redhat.com/support/policy/updates/rhscl) and for sure they will provide security patches (maybe even backports).
Even CentOS is [working on it](https://ci.centos.org/view/SCLo-pkgs-cbs/job/SCLo-pkg-rh-php72-rh-C7-candidate-x86_64/).
If I get it right, this should be the [testing repository](https://buildlogs.centos.org/centos/7/sclo/x86_64/sclo/sclo-php72/?C=M;O=D).

To recap, let's wait a little bit to see if upstream provides updated PHP packages, otherwise we can try to create the meta package to handle multiple PHP version.

> It should also be noted in the readme a recommendation to remove rh-php71 via yum 

I agree, let's remember to add such note.

> Lastly, is there any plans to create a v15 branch once NextCloud 16 is in production release?

Why not :) It's a matter to just create a branch, I will gladly do it.
But I can't assure I will build RPM on regular basis for branch 15 (we are already late for 15.0.7).

Maybe if someone want to join in, we could add Travis support for automatic builds and push created RPM to Packagecloud (or similar).

**Comment by chilinux** [comment]
NextCloud 16.0.0 stable has now appeared in the downloads directory.

Once the branch 15 is created, I will take responsibility for that version.

As far as RHEL maintaining the SCL PHP 7.2 packages, there is still no indication of them making progress on addressing the security CVEs for 2019 either by upgrading to a newer PHP 7.2.x or by backporting the fixes.  The ci.centos.org still shows in the April 24 build log they are still rebuilding the same 7.2.10-3 package they released in December 2018.  Again, this was created before the latest security disclosures.  The testing repository you linked to has even earlier 7.2.10-1 packages from November 2018 which also don't have fixes for the 2019 security issues.  None of this is showing progress in backporting the multiple security fixes in 7.2.14 from Feb 2019.

I'm not trying to be overly alarmist and not trying to indicate there is any immediate need to offer alternatives to SCL right away.  However, for the mainstream OS, I am used to seeing updates within 30 days of upstream security fixes and that isn't what I'm seeing for the maintenance of SCL.  I would still like to explore options to make it easier for end users to choose for themselves if RHEL/CentOS/SCL is not able to address this soon.

Thanks

**Comment by gsanchietti** [comment]
Sorry for the delayed answer, I was on vacation.

I've just created the `v15` branch, we will start working on 16 soon.

IMO, at first, we should run Nextcloud 16 on PHP 7.1 to be sure to avoid regressions, than we could try to work on an alternative to SCL. Let's see if others have the same requirements before starting the big refactor ;)

**Comment by chilinux** [comment]
NextCloud 16 system requirements recommend PHP 7.2 or PHP 7.3 with PHP 7.1 still listed as supported.

I'm guessing 7.1 is no longer listed as recommended because an 8 month support cycle for NextCloud 16 goes to the end of the year while PHP 7.1 goes EoL at the end of November.

NextCloud 17 is due out in August.  Would you be open to switch NextCloud 16 to 7.2 at that time?  Or would you like to consider jumping past 7.2 and just using 7.3 for both v16 and v17 around August?

Thanks

**Comment by gsanchietti** [comment]
AFAIK there is still no support for PHP 7.3 from [upstream](https://www.softwarecollections.org/en/scls/?search=php&page=1).

So I would jump straight to PHP 7.2 which will have a longer life-cycle. I already did some tests and NC seems running smoothly.

**Comment by gsanchietti** [comment]
@chilinux there is a new release ready to be tested: https://github.com/nextcloud/server-packages/releases/tag/centos-16.0.0 ;)

**Comment by gsanchietti** [comment]
Nextcloud 16 with PHP 7.2 has been released.
